### PR TITLE
Override `load_state_dict` for the GPT LightningModule

### LINKED ISSUE #795: RuntimeError: Error(s) in loading state_dict for LightningGPTModule
I am trying to use the "openwebtext_trainer.py" pretraining script on my custom data with tiny llama model. I trainer one epoch and then saved the last checkpoint.

I am trying to resume from the saved checkpoint but I am getting this error:

```
Using bfloat16 Automatic Mixed Precision (AMP)
GPU available: True (cuda), used: True
TPU available: False, using: 0 TPU cores
IPU available: False, using: 0 IPUs
HPU available: False, using: 0 HPUs
Seed set to 1337
{'model_name': 'tiny-llama-1.1b', 'name': 'lit-tiny-llama-1.1b', 'save_interval': 1000, 'eval_interval': 1000, 'eval_iters': 100, 'log_interval': 1, 'learning_rate': 0.0004, 'batch_size': 125, 'micro_batch_size': 5, 'gradient_accumulation_steps': 25, 'max_iters': 50, 'weight_decay': 0.1, 'beta1': 0.9, 'beta2': 0.95, 'decay_lr': True, 'warmup_iters': 2000, 'lr_decay_iters': 50, 'min_lr': 6e-05}
Loading model with {'name': 'tiny-llama-1.1b', 'hf_config': {'org': 'TinyLlama', 'name': 'TinyLlama-1.1B-intermediate-step-955k-token-2T'}, 'block_size': 2048, 'vocab_size': 32000, 'padding_multiple': 64, 'padded_vocab_size': 32000, 'n_layer': 22, 'n_head': 32, 'n_embd': 2048, 'rotary_percentage': 1.0, 'parallel_residual': False, 'bias': False, 'lm_head_bias': False, 'n_query_groups': 4, 'shared_attention_norm': False, '_norm_class': 'RMSNorm', 'norm_eps': 1e-05, '_mlp_class': 'LLaMAMLP', 'gelu_approximate': 'none', 'intermediate_size': 5632, 'rope_condense_ratio': 1, 'rope_base': 10000, 'head_size': 64, 'rope_n_elem': 64}
Time to instantiate model: 0.00 seconds.
Checkpoint directory /home/jovy/pretrain/out/lit-tiny-llama-1.1b exists and is not empty.
Restoring states from the checkpoint path at /home/jovy/pretrain/out/lit-tiny-llama-1.1b/last.ckpt
Traceback (most recent call last):
  File "lit-gpt/pretrain/openwebtext_trainer.py", line 206, in <module>
    CLI(main)
  File "/home/jovyan/.conda/envs/jupyter-base/lib/python3.8/site-packages/jsonargparse/_cli.py", line 96, in CLI
    return _run_component(components, cfg_init)
  File "/home/jovyan/.conda/envs/jupyter-base/lib/python3.8/site-packages/jsonargparse/_cli.py", line 181, in _run_component
    return component(**cfg)
  File "lit-gpt/pretrain/openwebtext_trainer.py", line 165, in main
    trainer.fit(model, train_dataloader, val_dataloader, ckpt_path="last")
  File "/home/jovy/.conda/envs/jupyter-base/lib/python3.8/site-packages/lightning/pytorch/trainer/trainer.py", line 543, in fit
    call._call_and_handle_interrupt(
  File "/home/jovy/.conda/envs/jupyter-base/lib/python3.8/site-packages/lightning/pytorch/trainer/call.py", line 44, in _call_and_handle_interrupt
    return trainer_fn(*args, **kwargs)
  File "/home/jovy/.conda/envs/jupyter-base/lib/python3.8/site-packages/lightning/pytorch/trainer/trainer.py", line 579, in _fit_impl
    self._run(model, ckpt_path=ckpt_path)
  File "/home/jovy/.conda/envs/jupyter-base/lib/python3.8/site-packages/lightning/pytorch/trainer/trainer.py", line 954, in _run
    self._checkpoint_connector._restore_modules_and_callbacks(ckpt_path)
  File "/home/jovy/.conda/envs/jupyter-base/lib/python3.8/site-packages/lightning/pytorch/trainer/connectors/checkpoint_connector.py", line 396, in _restore_modules_and_callbacks
    self.restore_model()
  File "/home/jovy/.conda/envs/jupyter-base/lib/python3.8/site-packages/lightning/pytorch/trainer/connectors/checkpoint_connector.py", line 276, in restore_model
    trainer.strategy.load_model_state_dict(self._loaded_checkpoint)
  File "/home/jovy/.conda/envs/jupyter-base/lib/python3.8/site-packages/lightning/pytorch/strategies/strategy.py", line 363, in load_model_state_dict
    self.lightning_module.load_state_dict(checkpoint["state_dict"])
  File "/home/jovy/.conda/envs/jupyter-base/lib/python3.8/site-packages/torch/nn/modules/module.py", line 2153, in load_state_dict
    raise RuntimeError('Error(s) in loading state_dict for {}:\n\t{}'.format(
RuntimeError: Error(s) in loading state_dict for LightningGPTModule:
	Unexpected key(s) in state_dict: "lm_head.weight", "transformer.wte.weight", "transformer.h.0.norm_1.weight", "transformer.h.0.attn.attn.weight", "transformer.h.0.attn.proj.weight", "transformer.h.0.norm_2.weight", "transformer.h.0.mlp.fc_1.weight", "transformer.h.0.mlp.fc_2.weight", "transformer.h.0.mlp.proj.weight", "transformer.h.1.norm_1.weight", "transformer.h.1.attn.attn.weight", "transformer.h.1.attn.proj.weight", "transformer.h.1.norm_2.weight", "transformer.h.1.mlp.fc_1.weight", "transformer.h.1.mlp.fc_2.weight", "transformer.h.1.mlp.proj.weight", "transformer.h.2.norm_1.weight", "transformer.h.2.attn.attn.weight", "transformer.h.2.attn.proj.weight", "transformer.h.2.norm_2.weight", "transformer.h.2.mlp.fc_1.weight", "transformer.h.2.mlp.fc_2.weight", "transformer.h.2.mlp.proj.weight", "transformer.h.3.norm_1.weight", "transformer.h.3.attn.attn.weight", "transformer.h.3.attn.proj.weight", "transformer.h.3.norm_2.weight", "transformer.h.3.mlp.fc_1.weight", "transformer.h.3.mlp.fc_2.weight", "transformer.h.3.mlp.proj.weight", "transformer.h.4.norm_1.weight", "transformer.h.4.attn.attn.weight", "transformer.h.4.attn.proj.weight", "transformer.h.4.norm_2.weight", "transformer.h.4.mlp.fc_1.weight", "transformer.h.4.mlp.fc_2.weight", "transformer.h.4.mlp.proj.weight", "transformer.h.5.norm_1.weight", "transformer.h.5.attn.attn.weight", "transformer.h.5.attn.proj.weight", "transformer.h.5.norm_2.weight", "transformer.h.5.mlp.fc_1.weight", "transformer.h.5.mlp.fc_2.weight", "transformer.h.5.mlp.proj.weight", "transformer.h.6.norm_1.weight", "transformer.h.6.attn.attn.weight", "transformer.h.6.attn.proj.weight", "transformer.h.6.norm_2.weight", "transformer.h.6.mlp.fc_1.weight", "transformer.h.6.mlp.fc_2.weight", "transformer.h.6.mlp.proj.weight", "transformer.h.7.norm_1.weight", "transformer.h.7.attn.attn.weight", "transformer.h.7.attn.proj.weight", "transformer.h.7.norm_2.weight", "transformer.h.7.mlp.fc_1.weight", "transformer.h.7.mlp.fc_2.weight", "transformer.h.7.mlp.proj.weight", "transformer.h.8.norm_1.weight", "transformer.h.8.attn.attn.weight", "transformer.h.8.attn.proj.weight", "transformer.h.8.norm_2.weight", "transformer.h.8.mlp.fc_1.weight", "transformer.h.8.mlp.fc_2.weight", "transformer.h.8.mlp.proj.weight", "transformer.h.9.norm_1.weight", "transformer.h.9.attn.attn.weight", "transformer.h.9.attn.proj.weight", "transformer.h.9.norm_2.weight", "transformer.h.9.mlp.fc_1.weight", "transformer.h.9.mlp.fc_2.weight", "transformer.h.9.mlp.proj.weight", "transformer.h.10.norm_1.weight", "transformer.h.10.attn.attn.weight", "transformer.h.10.attn.proj.weight", "transformer.h.10.norm_2.weight", "transformer.h.10.mlp.fc_1.weight", "transformer.h.10.mlp.fc_2.weight", "transformer.h.10.mlp.proj.weight", "transformer.h.11.norm_1.weight", "transformer.h.11.attn.attn.weight", "transformer.h.11.attn.proj.weight", "transformer.h.11.norm_2.weight", "transformer.h.11.mlp.fc_1.weight", "transformer.h.11.mlp.fc_2.weight", "transformer.h.11.mlp.proj.weight", "transformer.h.12.norm_1.weight", "transformer.h.12.attn.attn.weight", "transformer.h.12.attn.proj.weight", "transformer.h.12.norm_2.weight", "transformer.h.12.mlp.fc_1.weight", "transformer.h.12.mlp.fc_2.weight", "transformer.h.12.mlp.proj.weight", "transformer.h.13.norm_1.weight", "transformer.h.13.attn.attn.weight", "transformer.h.13.attn.proj.weight", "transformer.h.13.norm_2.weight", "transformer.h.13.mlp.fc_1.weight", "transformer.h.13.mlp.fc_2.weight", "transformer.h.13.mlp.proj.weight", "transformer.h.14.norm_1.weight", "transformer.h.14.attn.attn.weight", "transformer.h.14.attn.proj.weight", "transformer.h.14.norm_2.weight", "transformer.h.14.mlp.fc_1.weight", "transformer.h.14.mlp.fc_2.weight", "transformer.h.14.mlp.proj.weight", "transformer.h.15.norm_1.weight", "transformer.h.15.attn.attn.weight", "transformer.h.15.attn.proj.weight", "transformer.h.15.norm_2.weight", "transformer.h.15.mlp.fc_1.weight", "transformer.h.15.mlp.fc_2.weight", "transformer.h.15.mlp.proj.weight", "transformer.h.16.norm_1.weight", "transformer.h.16.attn.attn.weight", "transformer.h.16.attn.proj.weight", "transformer.h.16.norm_2.weight", "transformer.h.16.mlp.fc_1.weight", "transformer.h.16.mlp.fc_2.weight", "transformer.h.16.mlp.proj.weight", "transformer.h.17.norm_1.weight", "transformer.h.17.attn.attn.weight", "transformer.h.17.attn.proj.weight", "transformer.h.17.norm_2.weight", "transformer.h.17.mlp.fc_1.weight", "transformer.h.17.mlp.fc_2.weight", "transformer.h.17.mlp.proj.weight", "transformer.h.18.norm_1.weight", "transformer.h.18.attn.attn.weight", "transformer.h.18.attn.proj.weight", "transformer.h.18.norm_2.weight", "transformer.h.18.mlp.fc_1.weight", "transformer.h.18.mlp.fc_2.weight", "transformer.h.18.mlp.proj.weight", "transformer.h.19.norm_1.weight", "transformer.h.19.attn.attn.weight", "transformer.h.19.attn.proj.weight", "transformer.h.19.norm_2.weight", "transformer.h.19.mlp.fc_1.weight", "transformer.h.19.mlp.fc_2.weight", "transformer.h.19.mlp.proj.weight", "transformer.h.20.norm_1.weight", "transformer.h.20.attn.attn.weight", "transformer.h.20.attn.proj.weight", "transformer.h.20.norm_2.weight", "transformer.h.20.mlp.fc_1.weight", "transformer.h.20.mlp.fc_2.weight", "transformer.h.20.mlp.proj.weight", "transformer.h.21.norm_1.weight", "transformer.h.21.attn.attn.weight", "transformer.h.21.attn.proj.weight", "transformer.h.21.norm_2.weight", "transformer.h.21.mlp.fc_1.weight", "transformer.h.21.mlp.fc_2.weight", "transformer.h.21.mlp.proj.weight", "transformer.ln_f.weight". 
```

Any help on what's causing this issue?

### PR DESCRIPTION
**Pull request by carmocca**
Fixes #795

### CODE REVIEW

**Comment by awaelchli** [comment]
@carmocca The alternative would be to delete the state_dict method, make it save a normal Lightning checkpoint, and then extend #803 to handle Lightning Trainer checkpoints. This would standardize the workflow since we need that conversion step anyway.

**Comment by carmocca** [comment]
@awaelchli That is a good suggestion, but I don't think it's exclusive to what this PR is currently implementing.

Here I only work around the `module.` key but the checkpoints produced by the trainer will still have the Trainer checkpoint format (with pl version, `state_dict` key, ...). Your solution should help in this latter case which I don't plan to address in this PR

File: pretrain/openwebtext_trainer.py [hunk 1]
```diff
@@ -60,11 +60,17 @@ def configure_model(self) -> None:
         self.module.apply(self.module._init_weights)
 
     def configure_optimizers(self) -> torch.optim.Optimizer:
+        if self.module is None:
```
Comment (Andrei-Aksionov) [inline pretrain/openwebtext_trainer.py:63]: Shouldn't it be a decorator, since the code is exactly the same in all the methods below?

Comment (carmocca) [reply pretrain/openwebtext_trainer.py:63]: I'm following the KISS engineering principle here

Comment (Andrei-Aksionov) [reply pretrain/openwebtext_trainer.py:63]: Ok. Fair enough.

**Pull request merged**